src/ape_cache/query.py

```diff
@@ -0,0 +1,135 @@
+
+    @property
+    def engine(self):
+        return create_engine(f"sqlite:///{self.database_file}", pool_pre_ping=True)
+
+    def init_db(self):
+        if self.database_file.is_file():
+            raise QueryEngineError("Database has already been initialized")
+
+        models.Base.metadata.create_all(bind=self.engine)
+
+    def purge_db(self):
+        if not self.database_file.is_file():
+            # Add check here to show we have a file that exists
+            raise QueryEngineError("Database must be initialized")
+
+        self.database_file.unlink()
+
+    @singledispatchmethod
+    def estimate_query(self, query: QueryType) -> Optional[int]:  # type: ignore
+        return None  # can't handle this query
+
+    @estimate_query.register
+    def estimate_block_query(self, query: BlockQuery) -> Optional[int]:
+        try:
+            with self.engine.connect() as conn:
+                q = conn.execute(
+                    text(
+                        """
+                        SELECT COUNT(*)
+                        FROM blocks
+                        WHERE blocks.number >= :start_block
+                        AND blocks.number <= :stop_block
+                        AND blocks.number mod :step = 0
+                        """
+                    ),
+                    start_block=query.start_block,
+                    stop_block=query.stop_block,
+                    step=query.step,
+                )
+                number_of_rows = q.rowcount
+
+        except Exception as err:
+            # Note: If any error, skip the data from the cache and continue to
+            #       query from provider.
+            logger.debug(err)
+            number_of_rows = 0
+
+        # NOTE: Assume 200 msec to get data from database
+        time_to_get_cached_records = 200 if number_of_rows > 0 else None
+        # NOTE: Very loose estimate of 0.75ms per block
+        return time_to_get_cached_records
+
+    @singledispatchmethod
+    def perform_query(self, query: QueryType) -> pd.DataFrame:  # type: ignore
+        raise QueryEngineError(f"Cannot handle '{type(query)}'.")
+
+    @perform_query.register
+    def perform_block_query(self, query: BlockQuery) -> pd.DataFrame:
+        with self.engine.connect() as conn:
+            q = conn.execute(
+                text(
+                    """
+                    SELECT :columns
+                    FROM blocks
+                    WHERE blocks.number >= :start_block
+                    AND blocks.number <= :stop_block
+                    AND blocks.number mod :step = 0
+                    """
+                ),
+                columns=",".join(query.columns),
+                start_block=query.start_block,
+                stop_block=query.stop_block,
+                step=query.step,
+            )
+            return pd.DataFrame(columns=query.columns, data=q.fetchall())
+
+    def update_cache(self, query: QueryType, result: pd.DataFrame):
```

Comment:
Also add a condition to not cache if we're on the local network

Reply:
We currently cannot query the cache if we are on the local network. Do you want me to remove the raise if we aren't connected to a network?